### ISSUE TITLE
Autodiscover does not find repos when using string filter

### ISSUE THREAD

**Issue by twarner-sofi** [renovatebot/renovate#9639]
**What Renovate type, platform and version are you using?**

Self hosted 24.119.2 pointed at gitlab.com

**Describe the bug**

As of the merge of https://github.com/renovatebot/renovate/pull/9453 autodiscover now does not find repos that it previously did. For example, `"autodiscoverFilter": "twarner-sofi-test/**/*"` does not match `twarner-sofi-test/subgroup-1/renovate-test-3`. However, `"autodiscoverFilter": ["twarner-sofi-test/**/*"]` does.

**Relevant debug logs**

<details><summary>Click me to see logs</summary>

```
renovate-local    | DEBUG: Combined config
...
renovate-local    |          "autodiscoverFilter": "twarner-sofi-test/**/*",
....
renovate-local    | DEBUG: Autodiscovering GitLab repositories
renovate-local    | DEBUG: Discovered 1826 project(s)
renovate-local    | DEBUG: None of the discovered repositories matched the filter
renovate-local    | DEBUG: Renovate exiting
```

vs working

```
renovate-local    | DEBUG: Combined config
...
renovate-local    |          "autodiscoverFilter": ["twarner-sofi-test/**/*"],
....
renovate-local    | DEBUG: Autodiscovering GitLab repositories
renovate-local    | DEBUG: Discovered 1826 project(s)
renovate-local    |  INFO: Autodiscovered repositories
renovate-local    |        "length": 1,
renovate-local    |        "repositories": ["twarner-sofi-test/subgroup-1/renovate-test-3"]
```
</details>

**Have you created a minimal reproduction repository?**

Use config.js for self-hosted bot

```javascript
module.exports = {
    endpoint: 'https://gitlab.com/api/v4/',
    platform: 'gitlab',
    onboardingConfig: {
        extends: ['config:base'],
    },
    autodiscover: true,
    autodiscoverFilter: 'twarner-sofi-test/**/*'
};
```

With a gitlab access token.

The repo I expect it to find is https://gitlab.com/twarner-sofi-test/subgroup-1/renovate-test-3

**Additional context**

Most likely caused by https://github.com/renovatebot/renovate/pull/9453

I realize the new format is an array, but for backwards compatibility, I expect single string to still work (migrated to length 1 array). If single string is no longer supported, an error should be emitted.

If incompatible changes in self-hosted config in minor versions is expected, there should probably be a note in the release notes.

BTW love the new array filter, will be migrating to it eventually but would like to be able to keep up to date in the meantime. 🙂

**Comment by rarkins** [comment]
@viceice if this is a regression error then I think we need to roll back? The goal was that the strings get migrated to array for existing users.

**Comment by viceice** [comment]
Yes, it's a regression 😔

**Comment by rarkins** [comment]
Created #9641 to roll back yesterday's change until we work out what's wrong.

Possibly it only affects entries in `config.js` only?

**Comment by viceice** [comment]
I'm not sure what happens. We need to investigate that.

**Comment by vitoyucepi** [comment]
@viceice https://github.com/renovatebot/renovate/pull/9453/files#diff-95ef0c8b78ce37c0f940c4328d6e34a3b67b5a940731efb9b5ddfa0df11d827dR35
If `config.autodiscoverFilter` is a `string` then this will iterate over letters of the string. As far as I understand.

**Comment by viceice** [comment]
There should happen a migration from string to string array before, but it seems not to happen.

We need to debug / test this to find the root cause.

**Comment by l0nax** [comment]
@viceice It seems like that `massageConfig(…)` is never called, i.e. when using a "basic" config (no presets).
Or am I missing something?

**Comment by viceice** [comment]
@l0nax I'm not sure what you mean, we reverted all the changes, so this should be fixed now.

@twarner-sofi Do you still have issues? If not, please close this issue

**Comment by l0nax** [comment]
I know that change has been reverted. I'm trying to understand what's going wrong and why `massageConfig` doesn't migrate string to array for the config entry.

**Comment by viceice** [comment]
Ok, I don't know yet the cause, this seems to need some local debugging.

**Comment by viceice** [comment]
@l0nax Please continue discussion in #8763. Also feel free to open a new pr 🙃

**Comment by viceice** [comment]
fixed in [24.119.4](https://github.com/renovatebot/renovate/releases/tag/24.119.4)
